src/main/java/htsjdk/samtools/filter/IntervalKeepPairFilter.java

```diff
@@ -0,0 +1,121 @@
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package htsjdk.samtools.filter;
+
+import htsjdk.samtools.SAMFileHeader;
+import htsjdk.samtools.SAMRecord;
+import htsjdk.samtools.SAMUtils;
+import htsjdk.samtools.util.Interval;
+import htsjdk.samtools.util.IntervalUtil;
+import htsjdk.samtools.util.OverlapDetector;
+
+import java.util.Collection;
+import java.util.List;
+
+/**
+ * Filter out SAMRecords where neither record of a pair overlaps a given set of intervals.
+ * If one record of a pair overlaps the interval list, than both are kept.
+ * It is required that the SAMRecords are passed in coordinate order, and have non-null SAMFileHeaders.
+ *
+ * @author kbergin@broadinstitute.org
+ */
+public class IntervalKeepPairFilter implements SamRecordFilter {
+    private final OverlapDetector<Interval> intervalOverlapDetector;
+
+    /**
+     * Prepare to filter out SAMRecords that do not overlap the given list of intervals
+     * @param intervals
+     */
+    public IntervalKeepPairFilter(final List<Interval> intervals) {
+        this.intervalOverlapDetector = new OverlapDetector<>(0,0);
+        this.intervalOverlapDetector.addAll(intervals, intervals);
+    }
+
+    /**
+     * Determines whether a SAMRecord matches this filter
+     * Takes record, finds its mate. Checks if either record overlaps the current interval using overlap detector
+     * If yes, return false -> don't filter it out
+     *
+     * If a read is secondary or supplementary, filter read out.
+     * Use IntervalFilter if you want to keep these reads, but NOTE:
+     * the resulting bam will not be valid.
+     *
+     *
+     * @param record the SAMRecord to evaluate
+     * @return true if the SAMRecord matches the filter, otherwise false
+     */
+    public boolean filterOut(final SAMRecord record) {
+        if (record.getNotPrimaryAlignmentFlag() || record.getSupplementaryAlignmentFlag()) {
+           return true;
+        }
+
+        final boolean readOneOverlapsIntervalList;
+        if (record.getReadUnmappedFlag()) {
+            readOneOverlapsIntervalList = false;
+        } else {
+            readOneOverlapsIntervalList = hasOverlaps(record.getReferenceName(), record.getStart(), record.getEnd());
+        }
+
+        if (readOneOverlapsIntervalList) {
+            return false;
+        }
+
+        final boolean readTwoOverlapsIntervalList;
+        if (record.getMateUnmappedFlag()) {
+            readTwoOverlapsIntervalList = false;
+        } else {
+            readTwoOverlapsIntervalList = hasOverlaps(record.getMateReferenceName(), record.getMateAlignmentStart(), SAMUtils.getMateAlignmentEnd(record));
+        }
+
+        // By this point, readOneOverlapsIntervalList is always true
+        return !readTwoOverlapsIntervalList;
+    }
+
+    /**
+     * Get read interval and check for overlaps over intervals in list
```

Comment:
Grammar.  Perhaps `Returns true true if SAMRecord overlaps any intervals in list, false otherwise.`